flow/src/main/scala/org/alephium/flow/core/BlockFlowGroupView.scala

```diff
@@ -141,18 +127,65 @@ object BlockFlowGroupView {
   ) extends BlockFlowGroupView[WS] {
     def worldState: WS = _worldState
 
-    def getPreOutput(outputRef: TxOutputRef): IOResult[Option[TxOutput]] = {
+    def getPreContractOutput(outputRef: ContractOutputRef): IOResult[Option[ContractOutput]] = {
+      if (TxUtils.isSpent(blockCaches, outputRef)) {
+        Right(None)
+      } else {
+        worldState.getOutputOpt(outputRef).flatMap {
+          case Some(output: ContractOutput) =>
+            Right(Some(output))
+          case Some(_) =>
+            Left(WorldState.expectedContractError)
+          case None =>
+            val index = blockCaches.indexWhere(_.relatedOutputs.contains(outputRef))
+            if (index != -1) {
+              blockCaches(index).relatedOutputs.get(outputRef) match {
+                case Some(output: ContractOutput) =>
+                  Right(Some(output))
+                case Some(_) =>
+                  Left(WorldState.expectedContractError)
+                case None =>
+                  Right(None)
+              }
+            } else {
+              Right(None)
+            }
+        }
+      }
+    }
+
+    def getPreAssetOutputInfo(outputRef: AssetOutputRef): IOResult[Option[AssetOutputInfo]] = {
       if (TxUtils.isSpent(blockCaches, outputRef)) {
         Right(None)
       } else {
-        worldState.getOutputOpt(outputRef).map {
-          case Some(output) => Some(output)
+        worldState.getOutputOpt(outputRef).flatMap {
+          case Some(output: AssetOutput) =>
+            Right(
+              Some(AssetOutputInfo(outputRef, output, PersistedOutput))
+            )
+          case Some(_) =>
+            Left(WorldState.expectedAssetError)
           case None =>
             val index = blockCaches.indexWhere(_.relatedOutputs.contains(outputRef))
             if (index != -1) {
-              Some(blockCaches(index).relatedOutputs(outputRef))
+              blockCaches(index).relatedOutputs.get(outputRef) match {
+                case Some(output: AssetOutput) =>
+                  Right(
+                    Some(
+                      AssetOutputInfo(
+                        outputRef,
+                        output,
+                        UnpersistedBlockOutput
+                      )
+                    )
+                  )
+                case Some(_) =>
+                  Left(WorldState.expectedAssetError)
+                case None =>
+                  Right(None)
```

Comment:
@pragmaxim let's add a simple test to cover this branch. In fact, I am surprised this is not covered since we have so many tests.

Reply:
This one cannot happen @polarker as if `blockCaches(index).relatedOutputs.get(outputRef)` returned `None`, the condition above `blockCaches.indexWhere(_.relatedOutputs.contains(outputRef))` would return `-1` and this branch would not be executed, also the array would get illegal index anyway.

Reply:
After this chance I am quite sure that codecov has high chance for making a false positive as it behaves non-deterministicaly.